modin/experimental/engines/omnisci_on_native/io.py

```diff
@@ -244,14 +245,82 @@ def read_csv(
             )
             if not use_modin_impl:
                 raise ArrowEngineException(error_message)
+
+            # TODO: potentially do this only in a strict compatibility mode as this may
+            # really harm performance for wide frames
+            # if STRICT_COMPATIBILITY:
+            column_names = pandas.read_csv(
+                **dict(
+                    mykwargs,
+                    nrows=0,
+                    skipfooter=0,
+                    engine=None if eng == "arrow" else engine,
+                ),
+            ).columns.tolist()
+
             if isinstance(dtype, dict):
                 column_types = {c: cls._dtype_to_arrow(t) for c, t in dtype.items()}
             else:
-                column_types = cls._dtype_to_arrow(dtype)
+                column_types = (
+                    {c: cls._dtype_to_arrow(dtype) for c in column_names}
+                    if dtype is not None
+                    # if dtype is not specified let Arrow infer column dtypes
+                    else {}
+                )
+
+            # By default Arrow's read_csv tries to infer types for all of the columns,
+            # including date-time ones. Pandas on the opposite does not parse date
+            # columns unless it's explicitly requested via the 'parse_dates' parameter.
+            # Aligning these differences requires passing explicit type-scheme to Arrow
+            # which includes proper types for datetime-like columns.
+            if parse_dates is not None:
+                # Deducing columns which potentially have a timestamp data-type to prohibit for
+                # Arrow to infer their types by explicitly setting them to string.
+                # TODO: potentially do this only in a strict compatibility mode as this may
+                # really harm performance for wide frames
+                # if STRICT_COMPATIBILITY:
+                datetime_columns = (
+                    pandas.read_csv(
+                        **dict(
+                            mykwargs,
+                            nrows=1,
+                            parse_dates=column_names,
+                            dtype=None,
+                            engine=None if eng == "arrow" else engine,
+                        ),
+                    )
+                    .select_dtypes("datetime")
+                    .columns
+                )
 
-            if (type(parse_dates) is list) and type(column_types) is dict:
-                for c in parse_dates:
-                    column_types[c] = pa.timestamp("s")
+            if is_list_like(parse_dates):
+                parse_dates_labels = tuple(
+                    column_names[col] if isinstance(col, int) else col
+                    for col in parse_dates
+                )
+                for col in datetime_columns:
+                    if col not in parse_dates_labels:
+                        column_types[col] = pa.string()
```

Comment:
```suggestion
                for col in (frozenset(datetime_columns) - parse_dates_labels):
                    column_types[col] = pa.string()
```
this is also probably faster :) but needs checking

Reply:
It would be faster if `datetime_columns` initially was a `frozenset`, otherwise, conversion to set makes it slower

<details><summary>Measurements code</summary>

```python
import timeit

M = 64
N = 20

non_existing_values = tuple(f"non_existing_value{i}" for i in range(M))

def fn0(): # timeit() -> 23.84s
    values = tuple(f"col{i}" for i in range(N))
    b = 0
    for val in non_existing_values:
        if val not in values:
            b += 1

def fn1(): # timeit() -> 9.15s
    values = frozenset(f"col{i}" for i in range(N))
    b = 0
    for val in non_existing_values:
        if val not in values:
            b += 1

def fn2(): # timeit() -> 10.37s
    values = frozenset(f"col{i}" for i in range(N))
    b = 0
    for val in frozenset(non_existing_values) - values:
        b += 1

print(timeit.timeit(fn0))
print(timeit.timeit(fn1))
print(timeit.timeit(fn2))

```

</details>

Reply:
microbenchmarks to the rescue! 🙃